### PR TITLE
NIFI-1037 Created processor that handles HDFS' inotify event stream.

### PR DESCRIPTION
**Pull request by jjmeyer0**
NIFI-1037 provides a processor that allows a user to consume HDFS' notification event stream. The processor allows for a specific directory to be watched for specific events. Each event will be converted to a flow file where the content is the event information in JSON format.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        "to JSON and written to the flow file's content. For example, if event.type is APPEND then the content of the flow file will contain a JSON file containing the information about the " +
+        "append event. If successful the flow files are sent to the 'success' relationship. Be careful of where the generated flow files are stored. If the flow files are stored in one of " +
+        "processor's watch directories there will be a never ending flow of events. It is also important to be aware that this processor must consume all events. The filtering must happen within " +
+        "the processor. This is because the HDFS admin's event notifications API does not have filtering.")
+@Stateful(scopes = Scope.CLUSTER, description = "The last used transaction id is stored. This is used ")
+@SeeAlso({GetHDFS.class, FetchHDFS.class, PutHDFS.class, ListHDFS.class})
+public class GetHDFSEvents extends AbstractHadoopProcessor {
+    static final PropertyDescriptor POLL_DURATION = new PropertyDescriptor.Builder()
+            .name("Poll Duration")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: @jjmeyer0 we are trying to adopt a new convention by using both .name() and .displayName() to ensure backward compatibility. The thread discussion is available here [1]. Would you mind updating the PR accordingly?

[1] http://mail-archives.apache.org/mod_mbox/nifi-dev/201605.mbox/%3C5A6FDF1E-1889-46FE-A3C4-5D2F0A905979@apache.org%3E

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                    "Please see javadoc for org.apache.hadoop.hdfs.client.HdfsAdmin#getInotifyEventStream: {}", new Object[]{e});
+        }
+
+        updateClusterStateForTxId(stateManager);
+    }
+
+    private EventBatch getEventBatch(DFSInotifyEventInputStream eventStream, long duration, TimeUnit timeUnit) throws IOException, InterruptedException, MissingEventsException {
+        // According to the inotify API we should retry a few times if poll throws an IOException.
+        // Please see org.apache.hadoop.hdfs.DFSInotifyEventInputStream#poll for documentation.
+        int i = 0;
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: Would it make sense to have this "retry number" as a user-defined property? just asking.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        "This processor will poll for new events based on a defined duration. For each event received a new flow file will be created with the expected attributes and the event itself serialized " +
+        "to JSON and written to the flow file's content. For example, if event.type is APPEND then the content of the flow file will contain a JSON file containing the information about the " +
+        "append event. If successful the flow files are sent to the 'success' relationship. Be careful of where the generated flow files are stored. If the flow files are stored in one of " +
+        "processor's watch directories there will be a never ending flow of events. It is also important to be aware that this processor must consume all events. The filtering must happen within " +
+        "the processor. This is because the HDFS admin's event notifications API does not have filtering.")
+@Stateful(scopes = Scope.CLUSTER, description = "The last used transaction id is stored. This is used ")
+@SeeAlso({GetHDFS.class, FetchHDFS.class, PutHDFS.class, ListHDFS.class})
+public class GetHDFSEvents extends AbstractHadoopProcessor {
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: You will need to add the processor in the services file to make it visible in NiFi:
https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/resources/META-INF/services/org.apache.nifi.processor.Processor

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/EventAttributes.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.hadoop.inotify;
+
+final class EventAttributes {
+    static final String EVENT_PATH = "hdfs.inotify.event.path";
+    static final String EVENT_TYPE = "hdfs.inotify.event.type";
+    static final String MIME_TYPE = "mime.type";
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/EventAttributes.java]: I think you can use the existing Core attribute for mime type, don't you think?

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java (hunk 2 above)
Comment (jjmeyer0) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: Yes, after looking at the javadoc again it does say retry at least 3 times. I'll update this.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java (hunk 1 above)
Comment (jjmeyer0) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: Thanks for the link. I'll look into this and update accordingly.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/EventAttributes.java (hunk 4 above)
Comment (jjmeyer0) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/EventAttributes.java]: The reason I did this is so that I wouldn't have to hardcode a string in the WriteAttribute annotation and then use the CoreAttributes and EventAttributes when writing said attributes. I thought it would be simpler to consolidated into one class. If it is preferred to use CoreAttributes I can switch mime.type to do so.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                                public void process(OutputStream out) throws IOException {
+                                    out.write(OBJECT_MAPPER.writeValueAsBytes(e));
+                                }
+                            });
+                            flowFiles.add(flowFile);
+                        }
+                    }
+
+                    getLogger().debug("Transferring all event flow files.");
+                    session.transfer(flowFiles, REL_SUCCESS);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: It'd be nice to add the creation of a provenance event with something like `session.getProvenanceReporter().receive(...)` with some details.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        "to JSON and written to the flow file's content. For example, if event.type is APPEND then the content of the flow file will contain a JSON file containing the information about the " +
+        "append event. If successful the flow files are sent to the 'success' relationship. Be careful of where the generated flow files are stored. If the flow files are stored in one of " +
+        "processor's watch directories there will be a never ending flow of events. It is also important to be aware that this processor must consume all events. The filtering must happen within " +
+        "the processor. This is because the HDFS admin's event notifications API does not have filtering.")
+@Stateful(scopes = Scope.CLUSTER, description = "The last used transaction id is stored. This is used ")
+@SeeAlso({GetHDFS.class, FetchHDFS.class, PutHDFS.class, ListHDFS.class})
+public class GetHDFSEvents extends AbstractHadoopProcessor {
+    static final PropertyDescriptor POLL_DURATION = new PropertyDescriptor.Builder()
+            .name("Poll Duration")
+            .description("How long to poll for HDFS events.")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: I don't have specific suggestion but I would rephrase the description of this property. Maybe something talking about "batch".  I'd also mention the comment I see in the javadoc API: "The time before the method returns may exceed the specified amount of time by up to the time required for an RPC to the NameNode." Just to be sure the user understands that each time the processor is triggered, it waits at list this duration.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+                    getLogger().debug("Transferring all event flow files.");
+                    session.transfer(flowFiles, REL_SUCCESS);
+                }
+
+                lastTxId = eventBatch.getTxid();
+            }
+        } catch (IOException | InterruptedException e) {
+            getLogger().error("Unable to get notification information: {}", new Object[]{e});
+            throw new ProcessException(e);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java]: Don't we want to yield the processor in this case?

**Comment by pvillard31** [comment]
Once comments are addressed, I'll have another look and give it a try but it looks good to me. It's a nice feature!

**Comment by pvillard31** [comment]
I have played with it and it works great.
One remark: I am wondering if the 'HDFS_PATH_TO_WATCH' property should be improved:
- accept the expression language to handle time-stamped directory?
- accept regular expressions?
- accept comma-separated list of paths?
Since we are polling all HDFS events it could make sense to have the best filtering options

**Comment by jjmeyer0** [comment]
Sorry for the late response, this week has been a bit crazy. I updated the processor to add all the suggestions (except the filtering update you talked about). I will look into updating the way the processor does filtering, but I have a couple concerns. For example, the comma-separated list of paths won't always work. HDFS directories can technically have commas. Also, I couldn't find an example of how regular expressions work with the expression language. I'm a but curious how'll they'll interact with one another. I'll try to find some time to play around with it this weekend to see.

**Comment by pvillard31** [comment]
@jjmeyer0 Yes you're absolutely right, I was not suggesting to have the three options together but rather think about what could be done to give more flexibility. I think there is a processor that combines both regular expression and expression language but cannot remember which one. Anyway, I'll have another look on Sunday or next week but it looks good to me!

**Comment by pvillard31** [comment]
Your changes look good to me. Let me know if you want to have a look to enable EL/RegEx on HDFS_PATH_TO_WATCH property (ReplaceText processor seems to accept both regex and expression language on search value property). 
Regarding annotations and mime type, I just had a discussion with @apiri and the conclusion is it'd be preferable to keep consistency with what has been done so far.  If this attribute changes in the future, only the documentation would be slightly incorrect.
Once done, I think we are ready to merge.

**Comment by jjmeyer0** [comment]
Hey, it's taking a bit longer than expected. I didn't have much time over the weekend. I'm adding the filtering and writing more unit tests. I'll push the updates tomorrow.

**Comment by pvillard31** [comment]
@jjmeyer0 LGTM +1 tested with workflows, mvn build OK.
Merged in master, but would you mind opening a specific PR for 0.x branch (Java 7 compatible):
GetHDFSEvents.java:[208,78] error: local variable e is accessed from within inner class; needs to be declared final

**Comment by trixpan** [comment]
this commit seems to break NIFI compatibility with MapR
